packages/blade/src/components/BottomSheet/BottomSheet.tsx

```diff
@@ -0,0 +1,413 @@
+
+  /*
+      1. The content should not be scrollable on lower or middle snapPoints
+      2. If we reach the top snapPoint we make the content scrollable
+      3. scrolling down the content will work as usual
+      4. but if the scroll position is at top and then we drag down on the content body
+         the bottom-sheet will start the dragging and we will set the scroll to 'none'
+    */
+  const bind = useDrag(
+    ({
+      active,
+      last,
+      cancel,
+      tap,
+      movement: [_mx, my],
+      velocity: [_vx, vy],
+      lastOffset: [_, lastOffsetY],
+      down,
+      args: [{ isContentDragging = false } = {}] = [],
+    }) => {
+      setIsDragging(down);
+      const rawY = lastOffsetY - my;
+
+      const lowerSnapPoint = dimensions.height * snapPoints[0];
+      const upperSnapPoint = dimensions.height * snapPoints[snapPoints.length - 1];
+
+      // predictedY is used to create velocity driven swipe
+      // the faster you swipe the more distance you cover
+      // this enables users to reach upper & lower snappoint with a single swipe
+      const predictedDistance = my * (vy / 2);
+      const predictedY = Math.max(
+        lowerSnapPoint,
+        Math.min(upperSnapPoint, rawY - predictedDistance * 2),
+      );
+
+      let newY = rawY;
+
+      if (down) {
+        // Ensure that users aren't able to drag the sheet
+        // more than the upperSnapPoint or maximum height of the sheet
+        // this is basically a clamp() function but creates a nice rubberband effect
+        const dampening = 0.55;
+        const totalHeight = grabHandleHeight + headerHeight + footerHeight + contentHeight;
+        if (totalHeight < upperSnapPoint) {
+          newY = rubberbandIfOutOfBounds(rawY, 0, totalHeight, dampening);
+        } else {
+          newY = rubberbandIfOutOfBounds(rawY, 0, upperSnapPoint, dampening);
+        }
+      } else {
+        newY = predictedY;
+      }
+
+      const isPosAtUpperSnapPoint = newY >= upperSnapPoint;
+
+      if (isContentDragging) {
+        if (isPosAtUpperSnapPoint) {
+          newY = upperSnapPoint;
+        }
+
+        // keep the newY at upper snap point
+        // until the scrollable content is not at top
+        // and previously saved Y position is greater than or equal to upper snap point
+        // Note: how using newY won't work here since we need the previous value of the newY
+        // since we always keep updating the newY,
+        // this is cruicial in making the scroll feel natural
+        const isContentScrolledAtTop = scrollRef.current && scrollRef.current.scrollTop <= 0;
+        if (lastOffsetY === upperSnapPoint && !isContentScrolledAtTop) {
+          newY = upperSnapPoint;
+        }
+        preventScrollingRef.current = newY < upperSnapPoint;
+      }
+
+      const shouldClose = newY < lowerSnapPoint;
+      if (shouldClose) {
+        setIsDragging(false);
+        close();
+        cancel();
+        return;
+      }
+
+      if (last) {
```

Comment:
What is last?

Reply:
These are states coming from useDrag https://use-gesture.netlify.app/docs/state/#gesture-state-attributes

last indicates if the if it's the last event or not, when the even is last event that means user has ended the dragging process and we snap to the nearest snapping point